pkg/core/permissions/matcher.go

```diff
@@ -0,0 +1,64 @@
+package permissions
+
+import (
+	"context"
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	"github.com/Kong/kuma/pkg/core/resources/manager"
+	"github.com/Kong/kuma/pkg/core/resources/store"
+)
+
+type TrafficPermissionsMatcher struct {
+	ResourceManager manager.ResourceManager
+}
+
+func (m *TrafficPermissionsMatcher) Match(ctx context.Context, dataplane *mesh_core.DataplaneResource) (*mesh_core.TrafficPermissionResourceList, error) {
+	permissions := &mesh_core.TrafficPermissionResourceList{}
+	if err := m.ResourceManager.List(ctx, permissions, store.ListByMesh(dataplane.GetMeta().GetMesh())); err != nil {
+		return nil, err
+	}
+	return MatchDataplaneTrafficPermissions(&dataplane.Spec, permissions), nil
+}
+
+func MatchDataplaneTrafficPermissions(dataplane *mesh_proto.Dataplane, trafficPermissions *mesh_core.TrafficPermissionResourceList) *mesh_core.TrafficPermissionResourceList {
```

Comment:
We should return a list of matching `TrafficPermissions` for each listener individually.

Let's create an issue an fix it in a separate PR

Reply:
Created https://github.com/Kong/kuma/issues/222